spring-kafka/src/main/java/org/springframework/kafka/support/converter/BatchMessagingMessageConverter.java

```diff
@@ -106,6 +141,21 @@ public void setGenerateTimestamp(boolean generateTimestamp) {
 			offsets.add(record.offset());
 			timestampTypes.add(record.timestampType().name());
 			timestamps.add(record.timestamp());
+			if (this.headerMapper != null) {
+				Map<String, Object> converted = new HashMap<>();
+				this.headerMapper.toHeaders(record.headers(), converted);
+				convertedHeaders.add(converted);
+			}
+			else {
+				if (record.headers().iterator().hasNext() && logger.isDebugEnabled() && !logged) {
```

Comment:
Gary, how about do not have this `record.headers().iterator().hasNext()`?
Since we call `natives.add(record.headers())` anyway we might log unconditionally, too

WDYT?